torch/csrc/distributed/c10d/intra_node_comm.cpp

```diff
@@ -0,0 +1,628 @@
+    for (size_t j = 0; j < kMaxDevices; ++j) {
+      oss << nvlMesh[i][j] << " ";
+    }
+    oss << std::endl;
+  }
+  os << oss.str();
+  return os;
+}
+
+/**
+ * Query the nvlink connection among devices.
+ */
+static NvlMesh getNvlMesh(size_t worldSize) {
+  using namespace c10::cuda;
+
+  NvlMesh nvlMesh = {};
+#ifndef USE_ROCM
+  auto driverApi = DriverAPI::get();
+  if (driverApi == nullptr) {
+    return nvlMesh;
+  }
+
+  std::vector<nvmlDevice_t> devices(worldSize, 0);
+  std::unordered_map<std::string, size_t> busIdToIdx;
+  std::vector<size_t> switchLinkCount(worldSize, 0);
+
+  for (size_t idx = 0; idx < worldSize; ++idx) {
+    cudaDeviceProp prop{};
+    if (cudaGetDeviceProperties(&prop, idx) == cudaErrorInvalidDevice) {
+      continue;
+    }
+    char busId[80];
+    snprintf(
+        busId,
+        sizeof(busId),
+        NVML_DEVICE_PCI_BUS_ID_FMT,
+        prop.pciDomainID,
+        prop.pciBusID,
+        prop.pciDeviceID);
+    TORCH_CHECK(
+        driverApi->nvmlDeviceGetHandleByPciBusId_v2_(busId, &devices[idx]) ==
+        NVML_SUCCESS);
+    busIdToIdx.emplace(std::make_pair(busId, idx));
+  }
+
+  for (size_t idx = 0; idx < worldSize; ++idx) {
+    for (size_t link = 0; link < kMaxNvLinks; ++link) {
+      nvmlReturn_t ret;
+      nvmlIntNvLinkDeviceType_t deviceType;
+      ret = driverApi->nvmlDeviceGetNvLinkRemoteDeviceType_(
+          devices[idx], link, &deviceType);
+      if (ret != NVML_SUCCESS) {
+        break;
+      }
+      // Remote device is GPU
+      if (deviceType == NVML_NVLINK_DEVICE_TYPE_GPU) {
+        nvmlPciInfo_t pciInfo;
+        ret = driverApi->nvmlDeviceGetNvLinkRemotePciInfo_v2_(
+            devices[idx], link, &pciInfo);
+        if (ret != NVML_SUCCESS) {
+          break;
+        }
+        auto it = busIdToIdx.find(pciInfo.busId);
+        if (it != busIdToIdx.end()) {
+          if (idx != it->second) {
+            nvlMesh[idx][it->second] += 1;
+          }
+        }
+        // Remote device is NVSwitch
+      } else if (deviceType == NVML_NVLINK_DEVICE_TYPE_SWITCH) {
+        switchLinkCount[idx] += 1;
+      }
+    }
+  }
+  // Process NVSwitch connections. For simplicity, we assume all NVSwitches
+  // are interconnected.
+  for (size_t i = 0; i < worldSize; ++i) {
+    for (size_t j = 0; j < worldSize; ++j) {
+      if (i == j) {
+        continue;
+      }
+      nvlMesh[i][j] += std::min(switchLinkCount[i], switchLinkCount[j]);
+    }
+  }
+  return nvlMesh;
+#else
+  return nvlMesh;
+#endif
+}
+
+/**
+ * Determine if the devices form a hybrid cube mesh
+ * topology given a NvlMesh.
+ */
+static bool isHybridCubeMesh(const NvlMesh nvlMesh) {
+  std::array<size_t, kMaxDevices> numNeighbors = {};
+  for (size_t i = 0; i < kMaxDevices; ++i) {
+    for (size_t j = 0; j < kMaxDevices; ++j) {
+      if (nvlMesh[i][j] > 0) {
+        numNeighbors[i] += 1;
+      }
+    }
+  }
+  for (size_t i = 0; i < kMaxDevices; ++i) {
+    if (numNeighbors[i] != 4) {
```

Comment:
Out of curiosity, is this just a necessary condition, or is it also sufficient?

Reply:
Hmm, good catch. I believed it was correct until I spent more time thinking about it and found a counter-example. I haven't yet found a minimal sufficient proof to determine whether a topology is an HCM. However, I do have a piece of logic in the .cu file that can determine if the topology can benefit from the HCM allreduce algorithm, which I believe is sufficient. I'll revisit this part.